pkg/cmd/releasenotes/options.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2021 ScyllaDB
+
+package releasenotes
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sort"
+
+	"github.com/scylladb/scylla-operator/pkg/genericclioptions"
+	githubql "github.com/shurcooL/githubv4"
+	"golang.org/x/oauth2"
+)
+
+type GenerateOptions struct {
+	genericclioptions.IOStreams
+
+	OrganizationName string
+	RepositoryName   string
+	RepositoryPath   string
+
+	GithubToken string
+
+	ReleaseName string
+	StartRef    string
+	EndRef      string
+
+	ghClient *githubql.Client
+}
+
+func NewGitGenerateOptions(streams genericclioptions.IOStreams) *GenerateOptions {
+	return &GenerateOptions{
+		IOStreams:        streams,
+		OrganizationName: "scylladb",
+		RepositoryName:   "scylla-operator",
+		RepositoryPath:   ".",
+	}
+}
+
+func (o GenerateOptions) Validate() error {
+	if o.ReleaseName == "" {
+		return errors.New("release name cannot be empty")
+	}
+	if o.StartRef == "" {
+		return errors.New("start ref cannot be empty")
+	}
+	if o.EndRef == "" {
+		return errors.New("end ref cannot be empty")
+	}
+
+	return nil
+}
+
+func (o *GenerateOptions) Complete(ctx context.Context) error {
+	var httpClient *http.Client
+	if o.GithubToken != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: o.GithubToken},
+		)
+		httpClient = oauth2.NewClient(ctx, ts)
+	}
+
+	o.ghClient = githubql.NewClient(httpClient)
+
+	return nil
+}
+
+func (o *GenerateOptions) Run(ctx context.Context) error {
+	commits, err := listCommitsBetween(o.RepositoryPath, o.StartRef, o.EndRef)
+	if err != nil {
+		return err
+	}
+
+	pullRequestNumbers := make(map[int]struct{})
+	for _, c := range commits {
+		if !isMergeCommit(c) {
+			continue
+		}
+
+		n, ok := parsePullRequestNumber(c.Message)
+		if !ok {
+			continue
+		}
+
+		pullRequestNumbers[n] = struct{}{}
+	}
+
+	sort.Slice(commits, func(i, j int) bool {
+		return commits[i].Committer.When.Before(commits[j].Committer.When)
+	})
+	startDate, endDate := commits[0].Committer.When, commits[len(commits)-1].Committer.When
+
+	prs, err := listPullRequests(ctx, o.ghClient, o.OrganizationName, o.RepositoryName, startDate, endDate)
+	if err != nil {
+		return err
+	}
+
+	filteredPRs := prs[:0]
+	for _, pr := range prs {
+		if _, ok := pullRequestNumbers[int(pr.Number)]; ok {
```

Comment:
Would it be better to filter the other way around? Making a map from GH pull request and iterating through the ones parsed from commits and looking those up in the map?

If there was a case where you have parsed a PR number which wasn't listed from GH query, this wouldn't fail while it should have.

Reply:
good catch, thanks!